aot/src/main/java/com/dylibso/chicory/experimental/aot/AotMethods.java

```diff
@@ -64,8 +66,45 @@ public static void tableInit(
         OpcodeImpl.TABLE_INIT(instance, tableidx, elementidx, size, elemidx, offset);
     }
 
+    // This is an ugly hack to workaround a bug on some JVMs (Temurin 17-)
+    private static final boolean enableMemCopyWorkaround;
+
+    static {
+        String workaround = System.getProperty("chicory.enableMemCopyWorkaround");
+
+        if (workaround != null) {
+            enableMemCopyWorkaround = Boolean.parseBoolean(workaround);
+        } else {
+            enableMemCopyWorkaround = Runtime.version().feature() < 21;
+        }
+    }
+
     public static void memoryCopy(int destination, int offset, int size, Memory memory) {
-        memory.copy(destination, offset, size);
+        // up to Java 17 the bug happens on various platforms we need to be conservative
+        if (enableMemCopyWorkaround) {
+            notInlinableMemoryCopy(destination, offset, size, memory);
+        } else {
+            memory.copy(destination, offset, size);
+        }
+    }
+
+    // HACK: this is a whole trick to prevent incorrect inlining of the JVM
+    private static ReadWriteLock lock = new ReentrantReadWriteLock();
```

Comment:
with the println, is the lock still needed?

Reply:
The lock is not needed if we implement the workaround  using megamorphic callsites.  
see: https://github.com/andreaTP/chicory/pull/14

WIth this lock based implemenation, all JDKs < 21 will have a big performance hit since all memcopy operations across ALL instances will contend on this one lock.

The downside of that alternate workaround is the runtime will need to hold a temporary class that implements the workaround until the JVM implement a fix.  I tend to think that's better than reducing the performance experience.

Reply:
I'm open to go with your proposal @chirino , probably the deal breaker would be running a few benchmarks and comparing results

Reply:
I'll try to whip something up.

Reply:
Quick mention, do we need to copy over the compiled additional class? Probably yes

Reply:
> do we need to copy over the compiled additional class

The workaround class was placed in Runtime so it's available without emitting it as part of the code generation.

Reply:
Cool, thank you for taking the time to do the checks and clean things up!